File: benchmarks/tsb/bench_pctchange_df.ts
/**
 * Benchmark: pctChangeDataFrame — percentage change across DataFrame columns.
 * Outputs JSON: {"function": "pctchange_df", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { pctChangeDataFrame, DataFrame } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 50;

const df = new DataFrame({
  a: Array.from({ length: SIZE }, (_, i) => i * 1.1 + 1),
  b: Array.from({ length: SIZE }, (_, i) => i * 0.5 + 2),
  c: Array.from({ length: SIZE }, (_, i) => i * 2.3 + 3),
});

for (let i = 0; i < WARMUP; i++) {
  pctChangeDataFrame(df);
  pctChangeDataFrame(df, { periods: 3 });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  pctChangeDataFrame(df);
  pctChangeDataFrame(df, { periods: 3 });
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "pctchange_df",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_pctchange_df.py
"""Benchmark: DataFrame.pct_change — percentage change across DataFrame columns."""
import json, time
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 50

df = pd.DataFrame({
    "a": [i * 1.1 + 1 for i in range(SIZE)],
    "b": [i * 0.5 + 2 for i in range(SIZE)],
    "c": [i * 2.3 + 3 for i in range(SIZE)],
})

for _ in range(WARMUP):
    df.pct_change()
    df.pct_change(periods=3)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.pct_change()
    df.pct_change(periods=3)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function": "pctchange_df", "mean_ms": round(total_ms / ITERATIONS, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))
